### ISSUE TITLE
Postgres support

### ISSUE THREAD

**Issue by natechols** [broadinstitute/cromwell#4759]
My environment is already using Postgres so I have made the necessary changes to the Cromwell code - mostly adjusting the database migrations to deal with different conventions for case sensitivity.  We absolutely do not want to maintain our own fork and would like to contribute these changes back.  Aside from the obvious requirement of not breaking the existing MySQL and HSQLDB support, what do I need to do to make an acceptable pull request?

**Comment by geoffjentry** [comment]
@natechols I'd **love** to see postgres support. The big thing which has held us back in the past has been concern that modifying the liquibase scripts would break existing users (as there are "a or b" type forks lurking in there). If you are able to work around that issue then definitely have at it. If that seems impossible, let's chat further.

**Comment by natechols** [comment]
Hi @geoffjentry, got an example of a fork in the liquibase scripts?  One workaround that I'm already using in a couple of places is having a separate `changeSet` specific to postgres - for example `addAutoIncrement` does not really work the way it does in MySQL.  But mostly I'm just enforcing the existing table/column name case conventions (Postgres wants to convert them all to lower-case).

The main problem right now is the `IMPORTS_ZIP` column in `WORKFLOW_STORE_ENTRY` - the migrations are working but as soon as I try to run a workflow I get this:
```
ERROR:  column "IMPORTS_ZIP" is of type bytea but expression is of type bigint at character 335
```
Typing this as a `Blob` in Slick appears to be the root of the problem, but I'll keep poking at it.

**Comment by kshakir** [comment]
> got an example of a fork in the liquibase scripts?

> One workaround that I'm already using in a couple of places is having a separate changeSet specific to postgres

See this link: http://www.liquibase.org/2009/03/what-effects-changeset-checksums.html

Those "couple of places" are likely the "forks" @geoffjentry was referring to. Additional changesets are fine, but "adjusting the database migrations" will add additional setup and test criteria regarding the MD5s. Here are two examples:

https://github.com/broadinstitute/cromwell/blob/70cff69799f149191bdaec5d8878dd2a3d5202b7/database/migration/src/main/resources/changesets/sync_not_null_constraints.xml#L20-L36

https://github.com/broadinstitute/cromwell/blob/70cff69799f149191bdaec5d8878dd2a3d5202b7/database/migration/src/main/resources/changesets/lengthen_wdl_value.xml#L5-L15

At minimum for the former changelog I suspect that fixes for Postgres (and [MariaDB](https://github.com/broadinstitute/cromwell/issues/4618) 🤔) will probably change the MD5s. As the link at the top says, there are workarounds to update/ignore the MD5s. But those workarounds will need to be implemented and CI tested-- along w/ [Postgres support](https://docs.travis-ci.com/user/database-setup/#postgresql).

**Comment by natechols** [comment]
@kshakir This could get messy, given the number of migrations.  My understanding is that adding `dbms="!postgresql"` will NOT affect the MD5s; given that, and the fact that the Postgres database is going to be created from scratch initially, do you have a preference for 1) trying to make the current migrations work for Postgres too (without breaking the MD5s), or 2) make all existing migrations non-Postgres and add a single comprehensive Postgres-specific migration?  Frankly I'm starting to wonder if it would be easier for me to just write out every CREATE statement to generate the current tables, and leave everything else alone.  (Especially since I had to change the Scala migrations too, which I suspect affected the MD5s.)

**Comment by kshakir** [comment]
> I'm starting to wonder if it would be easier for me to just write out every CREATE statement to generate the current tables

I'd prefer to use liquibase syntax as much as possible, versus [custom crafted SQL](https://www.liquibase.org/documentation/changes/sql.html).

> do you have a preference for 1) trying to make the current migrations work for Postgres too (without breaking the MD5s), or 2) make all existing migrations non-Postgres and add a single comprehensive Postgres-specific migration?

Of the two, I think it would be fantastic if we could do "1)".  Minimum requirements are that existing MySQL users can startup cromwell w/o a liquibase error. Ultimately, if you can get updated changelogs that actually don't cause collisions with existing MD5s for those populated databases that's one avenue that might work. If not, and "2)" is uglier but doesn't break things for MySQL, then so be it.

Side note: I suspect the existing Java/Scala changelogs can be a no-op / skip, assuming that anyone using Postgres will not need to migrate data for those specific changes. I believe we skipped those Java/Scala migrations for the in-memory HSQLDB instances.

Also, you didn't ask, but in my dream world Cromwell would have changelogs that:
- Use liquibase syntax vs. sql as much as possible
- Work for a new database
- Work for all old/populated databases
- Work for HSQLDB + MySQL + PostgreSQL + MariaDB
- Can be updated to add other databases if/when our [Slick](http://slick.lightbend.com/doc/3.2.3/supported-databases.html) calls work or cromwell switches to another SQL adapter

To get to that last point I've wondered how one would best handle the liquibase MD5 issue in the future, either suppressing the warnings and / or resetting the MD5s as needed.

**TL;DR Try 1), but as long as populated MySQL databases still startup with cromwell you're good!**

**Comment by natechols** [comment]
Since it's been a month I thought I'd post an update.  Main items:
- In general I'm having to make a lot more changes to the Scala code than I expected due to queries being written in a way that Postgres doesn't like.  (This isn't a criticism, more of a heads-up.)  Nothing functional, just refactoring.
- The way `Blob` is handled in Slick+Postgres turns out to be a massive pain.  I'm not sure if Slick is lazy-loading these fields or I just don't understand how it works under the hood, but the workaround is that the blobs need to be accessed as part of a transaction, which involved some refactoring of downstream processing.
- Semi-related question: is there a reason why the entire contents of the `importsZip` need to be stored in the database?  This quickly leads to an enormous METADATA_ENTRY table - possibly because I have call caching turned on, I haven't checked whether this is the cause yet.
- The auto-incremented fields that are `Option[Long]` in the data model can't be handled the same way in Postgres; I haven't decided whether this is simply different database behavior or a bug somewhere.  Anyway I found a workaround for that too.
- I may have messed up and branched from `master` in my fork by mistake, and in any case I'm definitely out of sync with your `develop`.  Do you have a preferred workflow to bring my branch up to date, i.e. to minimize the mess in the Git history?  (Despite using Git daily I'm still not totally sure what "best practice" is.)

At this point I can at least run a workflow using Postgres, minus call caching.  I'm going to be focusing on completing and testing this in the next couple of weeks.

**Comment by natechols** [comment]
@kshakir I believe I have a working and minimalist solution, pending coworker review.  Expect a pull request sometime this coming week.

**Comment by natechols** [comment]
Not sure whose job it is to close this, but as far as I'm concerned this is done.
